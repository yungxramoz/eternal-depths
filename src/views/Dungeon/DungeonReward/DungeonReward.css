.reward-container {
  height: 100%;
  display: flex;
  align-items: center;
  flex-direction: column;
  justify-content: center;
  text-align: center;
}

.reward-selection-container {
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 1em;
}
