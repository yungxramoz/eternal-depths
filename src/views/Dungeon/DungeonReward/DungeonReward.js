import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import RewardButton from '../../../components/organisms/RewardButton/RewardButton'
import RpgSeparator from '../../../components/templates/RpgSeperator/RpgSeparator'
import {
  battleVictory,
  calculatedHpReward,
  characterEquipItem,
  characterRecoverHp,
} from '../../../store/game/gameSlice'
import { generateItem, getItemStats } from '../../../utils/item-generator'
import { RPGUI_ICON } from '../../../constants/rpgui-icon'
import RpgIcon from '../../../components/atoms/RpgIcon/RpgIcon'
import './DungeonReward.css'
import ItemReplaceModal from './Modal/ItemReplaceModal'

const DungeonReward = () => {
  const dispatch = useDispatch()
  const hpReward = useSelector(calculatedHpReward)
  const level = useSelector((state) => state.game.character.current.level)
  const currentItems = useSelector(
    (state) => state.game.character.current.items,
  )
  const [gearReward] = useState(generateItem({ level }))
  const [replaceModal, setReplaceModal] = useState(false)

  const selectGear = () => {
    if (currentItems[gearReward.type]) {
      setReplaceModal(true)
      return
    }
    dispatch(characterEquipItem(gearReward))
    dispatch(battleVictory())
  }

  const selectHpRecovery = () => {
    dispatch(characterRecoverHp(hpReward))
    dispatch(battleVictory())
  }

  return (
    <>
      <div className="reward-container">
        <h2>Encounter slayed!</h2>
        <RpgIcon icon={RPGUI_ICON.TREASURE} />
        <p>Choose your reward:</p>
        <div className="reward-selection-container">
          <RewardButton
            icon={gearReward.icon}
            title={gearReward.name}
            subtitle={gearReward.rarity}
            description={getItemStats(gearReward)}
            onClick={selectGear}
          ></RewardButton>
          <RpgSeparator golden />
          <RewardButton
            icon={RPGUI_ICON.FOOD}
            title="Food"
            description={`Recover ${hpReward} HP`}
            onClick={selectHpRecovery}
          ></RewardButton>
        </div>
      </div>
      {replaceModal && (
        <ItemReplaceModal
          newItem={gearReward}
          onClose={() => setReplaceModal(false)}
          onRewardSelected={() => dispatch(battleVictory())}
        />
      )}
    </>
  )
}

export default DungeonReward
