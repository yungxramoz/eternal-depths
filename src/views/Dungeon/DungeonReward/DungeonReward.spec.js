import { configureStore } from '@reduxjs/toolkit'
import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import gameReducer from '../../../store/game/gameSlice'
import DungeonReward from './DungeonReward'

jest.mock('../../../utils/item-generator', () => ({
  generateItem: () => ({
    icon: 'test-icon',
    name: 'Test Item',
    rarity: 'Common',
    stats: { strength: 10 },
  }),
}))

describe('RewardModal', () => {
  let store
  beforeEach(() => {
    store = configureStore({
      reducer: {
        game: gameReducer,
      },
    })
  })

  it('renders correctly', () => {
    render(
      <Provider store={store}>
        <DungeonReward />
      </Provider>,
    )
    expect(document.body).toMatchSnapshot('RewardModal')
  })

  it('dispatches equipItem action when gear reward button is clicked', () => {
    const dispatchSpy = jest.spyOn(store, 'dispatch')
    render(
      <Provider store={store}>
        <DungeonReward />
      </Provider>,
    )
    fireEvent.click(screen.getByText('Test Item'))
    expect(dispatchSpy).toHaveBeenCalledWith(
      expect.objectContaining({ type: 'game/characterEquipItem' }),
    )
    expect(dispatchSpy).toHaveBeenCalledWith(
      expect.objectContaining({ type: 'game/battleVictory' }),
    )
  })

  it('dispatches recoverHp action when HP reward button is clicked', () => {
    const dispatchSpy = jest.spyOn(store, 'dispatch')
    render(
      <Provider store={store}>
        <DungeonReward />
      </Provider>,
    )
    fireEvent.click(screen.getByText('Food'))
    expect(dispatchSpy).toHaveBeenCalledWith(
      expect.objectContaining({ type: 'game/characterRecoverHp' }),
    )
    expect(dispatchSpy).toHaveBeenCalledWith(
      expect.objectContaining({ type: 'game/battleVictory' }),
    )
  })
})
