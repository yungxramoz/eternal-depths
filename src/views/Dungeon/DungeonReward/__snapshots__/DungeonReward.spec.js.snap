// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DungeonReward renders correctly: RewardModal 1`] = `
<body>
  <div>
    <div
      class="reward-container"
    >
      <h2>
        Encounter slayed!
      </h2>
      <div
        class="rpgui-icon treasure "
      />
      <p>
        Choose your reward:
      </p>
      <div
        class="reward-selection-container"
      >
        <button
          class="golden-button reward-button"
        >
          <div
            class="rpgui-container container framed framed-golden-2 golden-container"
          >
            <div
              class="item-container"
            >
              <div
                class="rpgui-icon test-icon item-icon"
              />
              <div
                class="item-content"
              >
                <p
                  class="item-title"
                >
                  Test Item
                </p>
                <p
                  class="item-subtitle"
                >
                  <i>
                    Common
                  </i>
                </p>
                <p
                  class="item-description"
                >
                  Test stats
                </p>
              </div>
            </div>
          </div>
        </button>
        <hr
          class="separator golden"
        />
        <button
          class="golden-button reward-button"
        >
          <div
            class="rpgui-container container framed framed-golden-2 golden-container"
          >
            <div
              class="item-container"
            >
              <div
                class="rpgui-icon food item-icon"
              />
              <div
                class="item-content"
              >
                <p
                  class="item-title"
                >
                  Food
                </p>
                <p
                  class="item-subtitle"
                >
                  <i />
                </p>
                <p
                  class="item-description"
                >
                  Recover 5 HP
                </p>
              </div>
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</body>
`;
